event/datadeal/data_verification_event.go

```diff
@@ -1 +1,215 @@
+
+	msgVoteDataVerification, err := makeDataVerificationVote(
+		d.reactor.OracleAcc().GetAddress(),
+		dataHash,
+		dealID,
+		voteOption,
+		oraclePrivKey.Serialize(),
+	)
+	if err != nil {
+		return err
+	}
+
+	txBuilder := panacea.NewTxBuilder(*d.reactor.QueryClient())
+
+	txBytes, err := generateTxBytes(msgVoteDataVerification, d.reactor.OracleAcc().GetPrivKey(), d.reactor.Config(), txBuilder)
+	if err != nil {
+		return err
+	}
+
+	if err := broadcastTx(d.reactor.GRPCClient(), txBytes); err != nil {
+		return err
+	}
+
+	return nil
+}
+
+func (d DataVerificationEvent) decryptData(oraclePrivKey *btcec.PrivateKey, dataSale *types.DataSale, deal *types.Deal, encryptedDataBz []byte) ([]byte, error) {
+	sellerAcc, err := d.reactor.QueryClient().GetAccount(dataSale.SellerAddress)
+	if err != nil {
+		return nil, err
+	}
+
+	sellerPubKeyBytes := sellerAcc.GetPubKey().Bytes()
+	sellerPubKey, err := btcec.ParsePubKey(sellerPubKeyBytes, btcec.S256())
+	if err != nil {
+		return nil, err
+	}
+
+	decryptSharedKey := crypto.DeriveSharedKey(oraclePrivKey, sellerPubKey, crypto.KDFSHA256)
+
+	decryptedData, err := crypto.DecryptWithAES256(decryptSharedKey, deal.Nonce, encryptedDataBz)
+	if err != nil {
+		return nil, err
+	}
+	return decryptedData, nil
+}
+
+func (d DataVerificationEvent) compareDataHash(dataSale *types.DataSale, decryptedData []byte) bool {
+	decryptedDataHash := sha256.Sum256(decryptedData)
+	decryptedDataHashStr := hex.EncodeToString(decryptedDataHash[:])
+
+	return decryptedDataHashStr == dataSale.DataHash
+}
+
+func (d DataVerificationEvent) verifyDataSaleAndGetVoteOption(dataSale *types.DataSale, decryptedData []byte, dataSchema []string) oracletypes.VoteOption {
+	if !d.compareDataHash(dataSale, decryptedData) {
+		return oracletypes.VOTE_OPTION_NO
+	}
+
+	err := validation.ValidateJSONSchemata(decryptedData, dataSchema)
+	if err != nil {
+		log.Warnf("failed to verify data. error(%s)", err)
+		return oracletypes.VOTE_OPTION_NO
+	}
+
+	return oracletypes.VOTE_OPTION_YES
+}
+
+func makeDataVerificationVote(voterAddress, dataHash string, dealID uint64, voteOption oracletypes.VoteOption, oraclePrivKey []byte) (*types.MsgVoteDataVerification, error) {
+	dataVerificationVote := &types.DataVerificationVote{
+		VoterAddress: voterAddress,
+		DealId:       dealID,
+		DataHash:     dataHash,
+		VoteOption:   voteOption,
+	}
+
+	key := secp256k1.PrivKey{
+		Key: oraclePrivKey,
+	}
+
+	marshaledVerificationVote, err := dataVerificationVote.Marshal()
+	if err != nil {
+		return nil, err
+	}
+
+	sig, err := key.Sign(marshaledVerificationVote)
+	if err != nil {
+		return nil, err
+	}
+
+	msgVoteDataVerification := &types.MsgVoteDataVerification{
+		DataVerificationVote: dataVerificationVote,
+		Signature:            sig,
+	}
+
+	return msgVoteDataVerification, nil
+}
+
+// generateTxBytes generates transaction byte array.
+// TODO: generateTxBytes function will be refactored.
```

Comment:
Are you considering to make this function reusable?

Reply:
Yes, this function used in every events. So I'll refactor this function reusable.